sim/druid/druid.go

```diff
@@ -3,30 +3,201 @@ package druid
+
+	wrathSpell        core.SingleTargetDirectDamageSpell
+	wrathCastTemplate core.SingleTargetDirectDamageSpellTemplate
+
+	InsectSwarmSpell        core.DamageOverTimeSpell
+	insectSwarmCastTemplate core.DamageOverTimeSpellTemplate
+}
+
+type SelfBuffs struct {
+	Omen      bool
+	Innervate bool
 }
 
 func (druid *Druid) GetCharacter() *core.Character {
 	return &druid.Character
 }
 
 func (druid *Druid) AddRaidBuffs(raidBuffs *proto.RaidBuffs) {
-	// TODO: Use talents to check for imp gotw
 	raidBuffs.GiftOfTheWild = proto.TristateEffect_TristateEffectRegular
+	if druid.Talents.ImprovedMarkOfTheWild > 0 { // ya ya whatever
+		raidBuffs.GiftOfTheWild = proto.TristateEffect_TristateEffectImproved
+	}
 }
+
+const ravenGoddessItemID = 32387
+
 func (druid *Druid) AddPartyBuffs(partyBuffs *proto.PartyBuffs) {
-	//buffs.Moonkin = proto.TristateEffect_TristateEffectRegular
-	// check for idol of raven goddess equipped
+	if druid.Talents.MoonkinForm { // assume if you have moonkin talent you are using it.
+		partyBuffs.MoonkinAura = proto.TristateEffect_TristateEffectRegular
+		for _, e := range druid.Equip {
+			if e.ID == ravenGoddessItemID {
+				partyBuffs.MoonkinAura = proto.TristateEffect_TristateEffectImproved
+				break
+			}
+		}
+	}
+}
+
+func (druid *Druid) Init(sim *core.Simulation) {
+	druid.starfire8CastTemplate = druid.newStarfireTemplate(sim, 8)
+	druid.starfire6CastTemplate = druid.newStarfireTemplate(sim, 6)
+	druid.moonfireCastTemplate = druid.newMoonfireTemplate(sim)
+	druid.wrathCastTemplate = druid.newWrathTemplate(sim)
+	druid.insectSwarmCastTemplate = druid.newInsectSwarmTemplate(sim)
 }
 
 func (druid *Druid) Reset(newsim *core.Simulation) {
+	druid.moonfireCastTemplate.Apply(&druid.MoonfireSpell)
+	druid.Character.Reset(newsim)
 }
 
+func (druid *Druid) Advance(sim *core.Simulation, elapsedTime time.Duration) {
+	druid.Character.RegenManaMP5Only(sim, elapsedTime)
+	druid.Character.Advance(sim, elapsedTime)
+}
+
+var InnervateCD = core.NewCooldownID()
+var InnervateAuraID = core.NewAuraID()
+
+func (druid *Druid) TryInnervate(sim *core.Simulation) {
+	// Currently just activates innervate on self when own mana is <75%
+	// TODO: get a real recommendation when to use this.
+	if druid.SelfBuffs.Innervate && druid.GetRemainingCD(InnervateCD, sim.CurrentTime) == 0 {
+		if druid.GetStat(stats.Mana)/druid.MaxMana() < 0.75 {
+			oldRegen := druid.PsuedoStats.SpiritRegenRateCasting
+			druid.PsuedoStats.SpiritRegenRateCasting = 1.0
+			druid.PsuedoStats.ManaRegenMultiplier *= 3.0
+			druid.AddAura(sim, core.Aura{
+				ID:      InnervateAuraID,
+				Name:    "Innervate",
+				Expires: sim.CurrentTime + time.Second*20,
+				OnExpire: func(sim *core.Simulation) {
+					druid.PsuedoStats.SpiritRegenRateCasting = oldRegen
+					druid.PsuedoStats.ManaRegenMultiplier /= 3.0
+				},
+			})
+			cd := time.Minute * 6
+			if druid.HasAura(Malorne4PcAuraID) {
+				cd -= time.Second * 48
+			}
+			druid.SetCD(InnervateCD, cd)
```

Comment:
Innervate has a mana cost too (4% of base mana). We should definitely add a helper function character.BaseMana() as this is used for many other spells too (like shaman totems).

Also, having a mana cost makes me think this definitely triggers the GCD.

Reply:
added GCD trigger

Reply:
I assume you just didnt have time to add the mana for cost now, but it still needs that. Also need to return something to indicate that Act() should skip the rest of its logic instead of trying to cast a spell. Pretty sure the current code will panic when it tries to cast a starfire after using innervate in the same Act() call.